src/components/molecules/DateInput/YearBar/YearBar.tsx

```diff
@@ -0,0 +1,256 @@
+        ) {
+          setCurrentYear(value);
+        }
+      });
+    }
+  };
+
+  const clearScrollParameters = () => {
+    setScrollVelocity(0);
+    setScrollTime(1);
+  };
+
+  const clearTimeInterval = useCallback(() => {
+    clearInterval(timeInterval);
+  }, [timeInterval]);
+
+  const clearAccelerateInterval = useCallback(() => {
+    setMovementX(0);
+    clearInterval(accelerateInterval);
+  }, [accelerateInterval]);
+
+  // MouseEvents
+
+  const onMouseMove = (e: React.MouseEvent<HTMLDivElement>) => {
+    if (!isMouseDown) return;
+    e.preventDefault();
+    setMovementX(e.movementX / Constants.movementDenominator);
+    manageCurrentYear();
+    setScrollingState(true);
+    if (wrapperRef.current) {
+      const x = e.pageX - wrapperRef.current?.offsetLeft;
+      const walk = x - startX;
+      wrapperRef.current.scrollLeft = scrollPosition - walk;
+    }
+  };
+
+  const onMouseLeave = () => {
+    setIsMouseDown(false);
+    if (isMouseDown) {
+      setScrollingState(false);
+    }
+    manageCurrentYear();
+    handleScrolling();
+  };
+
+  // TouchEvents
+
+  const onTouchMove = (e: React.TouchEvent<HTMLDivElement>) => {
+    if (!isMouseDown) return;
+    setScrollPosition(e.changedTouches[0].pageX);
+    manageCurrentYear();
+    setScrollingState(true);
+  };
+
+  const onTouchStart = (e: React.TouchEvent<HTMLDivElement>) => {
+    setIsMouseDown(true);
+    clearScrollParameters();
+    clearInterval(timeInterval);
+    setScrollingState(false);
+    if (wrapperRef.current) {
+      setStartX(e.changedTouches[0].pageX - wrapperRef.current.offsetLeft);
+      setScrollPosition(wrapperRef.current.scrollLeft);
+    }
+  };
+  const onMouseDown = (e: React.MouseEvent<HTMLDivElement>) => {
+    setIsMouseDown(true);
+    clearScrollParameters();
+    setScrollingState(false);
+    clearInterval(timeInterval);
+    clearInterval(accelerateInterval);
+    if (wrapperRef.current) {
+      setStartX(e.pageX - wrapperRef.current.offsetLeft);
+      setScrollPosition(wrapperRef.current.scrollLeft);
+    }
+  };
+
+  const onMouseUp = () => {
+    setIsMouseDown(false);
+    if (!isMobileDevice() && (movementX >= 0.66 || movementX <= -0.66)) {
+      accelerateScroll();
+      setScrollingState(true);
+      setTimeInterval(
+        setInterval(() => {
+          setScrollTime((prev) => prev + 15);
+        }, 1)
+      );
+    } else {
+      handleScrolling();
+    }
+  };
+  const onTouchEnd = () => {
+    setIsMouseDown(false);
+    setScrollingState(true);
+    setTimeInterval(
+      setInterval(() => {
+        setScrollTime((prev) => prev + 15);
+      }, 1)
+    );
+  };
+
+  const onScroll = () => {
+    if (!isMouseDown) {
+      setScrollPosition(+`${wrapperRef.current?.scrollLeft}`);
+      manageCurrentYear();
+    }
+  };
+
+  const manageScrollVelocity = useCallback(() => {
+    setScrollVelocity(+(Constants.scrollRoad / scrollTime).toFixed(5));
+  }, [scrollTime]);
+
+  useEffect(() => {
```

Comment:
Think about improving these `useEffect` hooks. It looks like a long row to hoe to improve or test. Maybe we can find some `useSwipe`/`useCarousel` hook to decrease this code.